src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs

```diff
@@ -759,8 +759,17 @@ private void Parse()
                 @"((?<username>[^:@]+)(:(?<password>[^:@]*))?@)?" +
                 serversPattern + @"(/" + databasePattern + ")?/?" + optionsPattern + "$";
 
+            Match badPercentMatch = null;
```

Comment:
It occurs to me that the spec is concerned with `%` only in the userinfo portion of the connection string. So my proposal is thus: perhaps it would be best to migrate this logic into a new helper function. Then we could create a ` ValidateUserNameAndPassword(...)` that's structured very similarly to `ExtractUserNameAndPassword(..)`.  `ValidateUserNameAndPassword(...)`  could call this new helper function, first checking the username and then the password, and throw an exception if the new helper function returns false. We could call this validation function prior to calling `ExtractUserNameAndPassword(...)`. Or something along those lines...  What do you think?

Reply:
That makes sense. I can go ahead and implement that if other reviewers think it would make sense.

Reply:
Would it make sense to call `ValidateUserNameAndPassword(...)`after we've extracted username and password? Or potentially from `ExtractUserNameAndPassword(...)`? Otherwise, it seems as though we'd either need to continue validating the whole connection string, or we'd be duplicating some of the work that gets done in `ExtractUserNameAndPassword(...)`.

Reply:
Doesn't seem like there is any advantage to only validating part of the connection string.

We support % escape sequences everywhere (not just username/password), so we might as well validate the whole connection string.

Reply:
In that case, I think it makes sense to leave the validation in the `Parse(...)` function.